### PR TITLE
Type lowering and SILGen for imported ObjC async decls.

### PR DESCRIPTION
**Pull request by jckarter**
Type lowering and (todo) SILGen support for calling ObjC declarations imported as async.

### CODE REVIEW

File: lib/SIL/IR/SILFunctionType.cpp [hunk 1]
```diff
@@ -1738,17 +1773,34 @@ static bool isPseudogeneric(SILDeclRef c) {
...
-    AbstractionPattern origResultType, CanType substFormalResultType) {
+void updateResultTypeForForeignInfo(
+    const ForeignInfo &foreignInfo, CanGenericSignature genericSig,
+    AbstractionPattern &origResultType, CanType &substFormalResultType) {
+  // If there's no error or async convention, the return type is unchanged.
+  if (!foreignInfo.Async && !foreignInfo.Error) {
+    return;
+  }
+  
+  // A foreign async convention means our lowered return type is Void, since
```
Comment (DougGregor) [inline lib/SIL/IR/SILFunctionType.cpp:1784]: In the synchronous-error case I've been commenting on, you could end up with (e.g.) `BOOL` here, as dictated by the foreign error convention.

**Review by DougGregor — COMMENTED** [review]
A few comments about some minor expectation mismatches with what the Clang importer is doing, but this is looking great!

File: lib/SIL/IR/AbstractionPattern.cpp [hunk 2]
```diff
@@ -123,32 +124,40 @@ AbstractionPattern TypeConverter::getAbstractionPattern(EnumElementDecl *decl) {
...
+  // Foreign async convention takes precedence.
+  if (foreignAsync.hasValue()) {
+    return EncodedForeignInfo(EncodedForeignInfo::Async,
+                              foreignAsync->completionHandlerParamIndex(),
+                              foreignAsync->completionHandlerErrorParamIndex());
+  } else if (foreignError.hasValue()) {
+    return EncodedForeignInfo(EncodedForeignInfo::Error,
+                              foreignError->getErrorParameterIndex(),
                               foreignError->isErrorParameterReplacedWithVoid(),
                               foreignError->stripsResultOptionality());
```
[the remark this answers is not in the document]
Comment (jckarter) [reply lib/SIL/IR/AbstractionPattern.cpp:140]: For now, I can have it throw up a diagnostic if this case comes up. It shouldn't be too hard to add support for handling either a synchronous or async error once the basic support is implemented.

**Comment by jckarter** [comment]
@swift-ci Please test

File: lib/SIL/IR/AbstractionPattern.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (DougGregor) [reply lib/SIL/IR/AbstractionPattern.cpp:140]: In my own experiments, this construct is very, very rare.

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by DougGregor** [comment]
Those linux failures look completely unrelated:

```
<unknown>:0: error: compiled module was created by an older version of the compiler; rebuild 'Swift' and try again: /home/buildnode/jenkins/workspace/swift-PR-Linux@2/branch-main/buildbot_linux/swift-linux-x86_64/lib/swift_static/linux/Swift.swiftmodule/x86_64-unknown-linux-gnu.swiftmodule
```

Let's retry.

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Pull request merged**